examples/DGmethods/held_suarez_forcing.jl

```diff
@@ -0,0 +1,1017 @@
+      F[1, _ρv], F[2, _ρv], F[3, _ρv] = u * ρv     , v * ρv + P, w * ρv
+      F[1, _ρw], F[2, _ρw], F[3, _ρw] = u * ρw     , v * ρw    , w * ρw + P
+    end
+    F[1, _dρe], F[2, _dρe], F[3, _dρe] = u * (ρe + P), v * (ρe + P), w * (ρe + P)
+
+    # #Derivative of T and Q:
+    # vTx, vTy, vTz = VF[_Tx], VF[_Ty], VF[_Tz]
+    # vρx, vρy, vρz = VF[_Tx], VF[_Ty], VF[_Tz]
+
+    # #Richardson contribution:
+    # SijSij = VF[_SijSij]
+
+    # #Dynamic eddy viscosity from Smagorinsky:
+    # Δsqr = aux[_a_Δsqr]
+    # (ν_e, D_e) = standard_smagorinsky(SijSij, Δsqr)
+    # # FIXME f_R = buoyancy_correction(SijSij, ρ, vρy)
+    # f_R = 1
+
+    # # Multiply stress tensor by viscosity coefficient:
+    # τ11, τ22, τ33 = VF[_τ11] * ν_e, VF[_τ22]* ν_e, VF[_τ33] * ν_e
+    # τ12 = τ21 = VF[_τ12] * ν_e
+    # τ13 = τ31 = VF[_τ13] * ν_e
+    # τ23 = τ32 = VF[_τ23] * ν_e
+
+    # # Viscous velocity flux (i.e. F^visc_u in Giraldo Restelli 2008)
+    # F[1, _ρu] -= τ11 * f_R ; F[2, _ρu] -= τ12 * f_R ; F[3, _ρu] -= τ13 * f_R
+    # F[1, _ρv] -= τ21 * f_R ; F[2, _ρv] -= τ22 * f_R ; F[3, _ρv] -= τ23 * f_R
+    # F[1, _ρw] -= τ31 * f_R ; F[2, _ρw] -= τ32 * f_R ; F[3, _ρw] -= τ33 * f_R
+
+    # # Viscous Energy flux (i.e. F^visc_e in Giraldo Restelli 2008)
+    # F[1, _dρe] -= u * τ11 + v * τ12 + w * τ13 + ν_e * k_μ * vTx
+    # F[2, _dρe] -= u * τ21 + v * τ22 + w * τ23 + ν_e * k_μ * vTy
+    # F[3, _dρe] -= u * τ31 + v * τ32 + w * τ33 + ν_e * k_μ * vTz
+  end
+end
+#md nothing # hide
+
+# FXG: Source function => Define the geopotential and Held-Suarez source from the solution and auxiliary variables
+function source!(S, Q, aux, t)
+  @inbounds begin
+
+    ## Store values
+    dρ, ρu, ρv, ρw, dρe = Q[_dρ], Q[_ρu], Q[_ρv], Q[_ρw], Q[_dρe]
+    ρ_ref, ρe_ref = aux[_a_ρ_ref], aux[_a_ρe_ref]
+    ϕ, ϕx, ϕy, ϕz = aux[_a_ϕ], aux[_a_ϕx], aux[_a_ϕy], aux[_a_ϕz]
+    x, y, z = aux[_a_x], aux[_a_y], aux[_a_z]
+
+    ## Add Geopotential source
+    S[_dρ ] = 0
+    if PDE_level_hydrostatic_balance
+      S[_ρu ] = -dρ * ϕx
+      S[_ρv ] = -dρ * ϕy
+      S[_ρw ] = -dρ * ϕz
+    else
+      ρ = ρ_ref + dρ
+      S[_ρu ] = -ρ * ϕx
+      S[_ρv ] = -ρ * ϕy
+      S[_ρw ] = -ρ * ϕz
+    end
+    S[_dρe] = 0
+    
+    # Coriolis force
+    if use_coriolis
+      coriolis_x =  2Omega * ρv
+      coriolis_y = -2Omega * ρu
+      coriolis_z =  0
+
+      S[_ρu ] += coriolis_x
+      S[_ρv ] += coriolis_y
+      S[_ρw ] += coriolis_z
+    end
+
+    ## Add Held-Suarez Source
+    #Extract Temperature
+    ρ = ρ_ref + dρ
+    ρe = ρe_ref + dρe
+    e = ρe / ρ
+    u, v, w = ρu / ρ, ρv / ρ, ρw / ρ
+    e_int = e - (u^2 + v^2 + w^2)/2 - ϕ
+    T = air_temperature(e_int)
+    P = air_pressure(T, ρ)
+    
+    #Create Held-Suarez forcing
+    (kv,kt,T_eq)=held_suarez_forcing(x,y,z,T,P,t) 
+    
+    #Apply forcing
+    if use_held_suarez_forcings
+      S[_ρu ]  -= kv*ρu
+      S[_ρv ]  -= kv*ρv
+      S[_ρw ]  -= kv*ρw
+      S[_dρe ] -= ( kt*ρ*cv_d*T + kv*(ρu*ρu + ρv*ρv + ρw*ρw)/ρ )
```

Comment:
Note, here w is the cartesian component of the velocity, not the radial velocity

Reply:
Ah, thanks, I had missed that. It remains true, however, that the vertical velocity component (however you get it here) should not appear in the kinetic energy for a consistent hydrostatic configuration.

Reply:
And if w is not vertical velocity, then why is the w component of the Coriolis acceleration set to zero? Only the vertical component is set to zero in the thin-shell approximation (which you do not need to make).

Reply:
Note that the `PDE_level_hydrostatic_balance` flag means that the reference state
in an exact hydrostatic balance and therefore some terms cancel out in the governing PDEs.
It does not mean that we are solving hydrostatic equations. Perhaps the name is a bit unfortunate.

Reply:
Ok, thanks. And is the w Coriolis component then in fact the vertical?

Reply:
The `w` Coriolis component is not in the vertical. We are solving equations for Cartesian velocity components. If Earth's axis of rotation is assumed to be along `z` then there is no Coriolis force in this direction. This is why this component is zero.

Reply:
Ok, makes sense. It’s pretty confusing to use the conventional u, v, w (usually eastward, northward, upward) for velocities in these Cartesian coordinates. We should change names to avoid that confusion.